File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/container/slots/MenuStorageSlot.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by FernFlower decompiler)
//

package generations.gg.generations.core.generationscore.common.world.container.slots;

import earth.terrarium.common_storage_lib.resources.item.ItemResource;
import earth.terrarium.common_storage_lib.storage.base.CommonStorage;
import earth.terrarium.common_storage_lib.storage.base.StorageSlot;
import earth.terrarium.common_storage_lib.storage.base.UpdateManager;
import earth.terrarium.common_storage_lib.storage.util.ModifiableItemSlot;
import net.minecraft.world.Container;
import net.minecraft.world.SimpleContainer;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;

public class MenuStorageSlot extends Slot {
    public static final Container EMPTY = new SimpleContainer(0);
    private final StorageSlot<ItemResource> storageSlot;
    private final ModifiableItemSlot modifiable;

    private ItemStack cached = null;

    public MenuStorageSlot(CommonStorage<ItemResource> storage, int slotIndex, int x, int y) {
        super(EMPTY, slotIndex, x, y);
        this.storageSlot = storage.get(slotIndex);
        StorageSlot var6 = this.storageSlot;
        if (var6 instanceof ModifiableItemSlot slot) {
            this.modifiable = slot;
            cached = modifiable.toItemStack();
        } else {
            throw new UnsupportedOperationException("Cannot create MenuStorageSlot from non-modifiable slot");
        }
    }

    public boolean mayPlace(ItemStack stack) {
        return !stack.isEmpty() && this.storageSlot.isResourceValid(ItemResource.of(stack));
    }

    public @NotNull ItemStack getItem() {
        return cached;
    }

    public void set(ItemStack stack) {
        this.modifiable.set(stack);
        cached = stack;
    }

    public void onQuickCraft(ItemStack oldStackIn, ItemStack newStackIn) {
    }

    public int getMaxStackSize() {
        return (int)this.storageSlot.getLimit((ItemResource)this.storageSlot.getResource());
    }

    public int getMaxStackSize(ItemStack stack) {
        return this.modifiable.getMaxAllowed(ItemResource.of(stack));
    }

    public boolean mayPickup(Player playerIn) {
        return !this.modifiable.isEmpty();
    }

    public @NotNull ItemStack remove(int amount) {
        ItemResource resource = (ItemResource)this.storageSlot.getResource();
        long extract = this.storageSlot.extract(resource, (long)amount, false);
        var itemStack = extract > 0L ? resource.toStack((int)extract) : ItemStack.EMPTY;

        cached = modifiable.toItemStack();

        return itemStack;

    }

    public void setChanged() {
        super.setChanged();
        UpdateManager.batch(new Object[]{this.storageSlot});
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/world/container/slots/MenuStorageSlot.kt
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by FernFlower decompiler)
//
package generations.gg.generations.core.generationscore.common.world.container.slots

import earth.terrarium.common_storage_lib.resources.item.ItemResource
import earth.terrarium.common_storage_lib.storage.base.CommonStorage
import earth.terrarium.common_storage_lib.storage.base.UpdateManager
import earth.terrarium.common_storage_lib.storage.util.ModifiableItemSlot
import net.minecraft.world.Container
import net.minecraft.world.SimpleContainer
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.ItemStack

open class MenuStorageSlot(private val storage: CommonStorage<ItemResource>, slotIndex: Int, x: Int, y: Int) :
    Slot(EMPTY, slotIndex, x, y) {
    private val storageSlot =
        storage[slotIndex]
    private var modifiable: ModifiableItemSlot

    init {
        if (storageSlot is ModifiableItemSlot) {
            this.modifiable = storageSlot
        } else {
            throw UnsupportedOperationException("Cannot create MenuStorageSlot from non-modifiable slot")
        }
    }

    override fun mayPlace(stack: ItemStack): Boolean {
        return !stack.isEmpty && storageSlot.isResourceValid(ItemResource.of(stack))
    }

    override fun getItem(): ItemStack {
        return storageSlot.resource.toStack(storageSlot.amount.toInt())
    }

    override fun set(stack: ItemStack) {
        modifiable.set(stack)
    }

    override fun onQuickCraft(oldStackIn: ItemStack, newStackIn: ItemStack) {
    }

    override fun getMaxStackSize(): Int {
        return storageSlot.getLimit(storageSlot.resource as ItemResource).toInt()
    }

    override fun getMaxStackSize(stack: ItemStack): Int {
        return modifiable.getMaxAllowed(ItemResource.of(stack))
    }

    override fun mayPickup(playerIn: Player): Boolean {
        return !modifiable.isEmpty
    }

    override fun remove(amount: Int): ItemStack {
        val resource = storageSlot.resource as ItemResource
        val extract = storageSlot.extract(resource, amount.toLong(), false)
        val itemStack = if (extract > 0L) resource.toStack(extract.toInt()) else ItemStack.EMPTY
        return itemStack
    }

    override fun setChanged() {
        super.setChanged()
        UpdateManager.batch(this.storageSlot)
    }

    companion object {
        val EMPTY: Container = SimpleContainer(0)
    }
}
